File: concurrency-async/ConcurrencyAsync.cpp
#include <chrono>
#include <future>
#include <iostream>
#include <string>
#include <thread>
#include <vector>

// --- Threads ---

void workerThread(const std::string& name) {
    std::this_thread::sleep_for(std::chrono::milliseconds(50));
    std::cout << "Thread '" << name << "' done\n";
}

void runThreads() {
    std::cout << "\n-- Threads --\n";
    std::thread t1(workerThread, "A");
    std::thread t2(workerThread, "B");
    t1.join();  // wait for t1 to finish
    t2.join();  // wait for t2 to finish
}

// --- std::async / std::future (async/await equivalent) ---

std::string fetchData(const std::string& name, int delayMs) {
    std::this_thread::sleep_for(std::chrono::milliseconds(delayMs));
    return "Data from '" + name + "'";
}

void runAsync() {
    std::cout << "\n-- std::async + std::future --\n";

    // Launch async tasks (may run on thread pool)
    std::future<std::string> fa = std::async(std::launch::async, fetchData, "source-A", 200);
    std::future<std::string> fb = std::async(std::launch::async, fetchData, "source-B", 100);
    std::future<std::string> fc = std::async(std::launch::async, fetchData, "source-C", 50);

    // .get() blocks until result is ready — like await
    std::cout << fa.get() << "\n";
    std::cout << fb.get() << "\n";
    std::cout << fc.get() << "\n";
}

// --- Parallel CPU work with multiple threads ---

void runParallel() {
    std::cout << "\n-- Parallel CPU work --\n";
    const int numThreads = 4;
    std::vector<std::thread> threads;
    std::vector<long long> results(numThreads, 0);

    for (int i = 0; i < numThreads; ++i) {
        threads.emplace_back([i, &results]() {
            long long sum = 0;
            for (int j = i * 250000; j < (i + 1) * 250000; ++j) sum += j;
            results[i] = sum;
        });
    }
    for (auto& t : threads) t.join();

    long long total = 0;
    for (auto r : results) total += r;
    std::cout << "Parallel sum (0..999999): " << total << "\n";
}

int main() {
    std::cout << "Language: C++\n";
    runThreads();
    runAsync();
    runParallel();
    return 0;
}

File: concurrency-async/ConcurrencyAsync.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

class ConcurrencyAsync
{
    // --- Threads ---
    static void RunThreads()
    {
        Console.WriteLine("\n-- Threads --");
        var t1 = new Thread(() => Console.WriteLine("Thread A running"));
        var t2 = new Thread(() => Console.WriteLine("Thread B running"));
        t1.Start();
        t2.Start();
        t1.Join();
        t2.Join();
    }

    // --- async/await ---
    static async Task<string> FetchDataAsync(string name, int delayMs)
    {
        await Task.Delay(delayMs);   // non-blocking wait
        return $"Data from '{name}'";
    }

    static async Task RunAsyncAwait()
    {
        Console.WriteLine("\n-- async/await --");

        // Sequential
        string a = await FetchDataAsync("source-A", 100);
        string b = await FetchDataAsync("source-B", 100);
        Console.WriteLine($"Sequential: {a} | {b}");

        // Parallel: Task.WhenAll runs tasks concurrently
        string[] results = await Task.WhenAll(
            FetchDataAsync("source-X", 200),
            FetchDataAsync("source-Y", 100),
            FetchDataAsync("source-Z", 50)
        );
        Console.WriteLine("Task.WhenAll: " + string.Join(", ", results));
    }

    // --- Parallel execution (CPU-bound work) ---
    static void RunParallel()
    {
        Console.WriteLine("\n-- Parallel (CPU-bound) --");

        // Parallel.For uses thread pool for data parallelism
        int[] squares = new int[5];
        Parallel.For(0, 5, i => squares[i] = i * i);
        Console.WriteLine("Parallel.For squares: " + string.Join(", ", squares));

        // Task.Run offloads CPU work off the calling thread
        var task = Task.Run(() =>
        {
            long sum = 0;
            for (int i = 0; i < 1_000_000; i++) sum += i;
            return sum;
        });
        Console.WriteLine("Task.Run sum: " + task.Result);
    }

    static async Task Main()
    {
        Console.WriteLine("Language: C#");
        RunThreads();
        await RunAsyncAwait();
        RunParallel();
    }
}
